Fix: Replacements can cause runtime assertions to disappear and can cause invalid inductor code.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153661

Lets explore firs a couple of problem related to replacements and runtime assertions.

#### example problem 1
if we have a runtime assertions that u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added, the function f(u0, s0) will become f(s0, s0), this leads to the assert  not being inserted during insert_deferred_runtime_asserts. 
The reason is that insert_deferred_runtime_asserts logic insert each assertion once all its inputs are seen,  but u0 will never be seen. Same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

#### example problem 2
Consider u0==s0, where u0 is coming from a call to .item() Imagine later on that a specialization happens to s0 to become 2. In that case s0 as input wont be seen during insert_deferred_runtime_asserts and the assertion won't be inserted in the graph. Worse, Inductor will generate some code that refers to s0 in the cpp wrapper while it does not exist, causing a failure.
internal xref: https://fb.workplace.com/groups/1075192433118967/permalink/1669766396994898/

## The solution :
Runtime assertions insertion loops depend on detecting that the symbols that are used in the runtime assertions are seen, note that those symbols are either graph inputs or generated in the graph from data dependent ops like .item().


The issues above happen when symbols are graph inputs, in order to force the symbols to exist in the graph and to be seen by the runtime assertions we do not do replacements on placeholders expressions during codegen and during runtime assertions insertion.

This should not have performance overhead, since we already optimized the graph with replacements, the only effect is not mistakenly dropping graph inputs that are used in runtime assertions.
I added extended testing. A solo unrelated follow up that I noticed, is that we might want to rename unbacked symbols in runtime assertions when we do unbacked renaming, but that's a different issue.

Other approaches that did not work :
#### ban replacements on unbacked.
1. does not work when we defer runtime assertions on backed ex: s0==s1. we could also ban such replacements
but problem 2 becomes more problematic.
2. Problem two, it affects the quality of reasoning ! in a bad way.

#### Apply specialization on runtime assertions before codegen .
1. Can fix some issues, but may lead also to runtime assertions becoming NOPs.
2. Does not fix the issue if not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected.



cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
fix specialized symbols after runtime assertions added

[ghstack-poisoned]

laithsakka commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5941,11 +5954,22 @@ def resimplify_floor_div(axioms: dict[sympy.Expr, sympy.Expr]) -> None:
+
             r = self._find(s)
+
+            if only_specialized_backed and not (r.is_integer or r.is_real):
```
if rhs is not integer, its guaranteed the input it still in the graph. 
the replacement is not needed for the runtime assert symbol. it could be that its ok if we relax this,idk if we would hit issues.

laithsakka replied:
actually this should be int float and not is_integer, is_real

laithsakka committed:
Update on "Fix: specializing symbols after runtime assertions added cause codegen issue"


Fix https://github.com/pytorch/pytorch/issues/153650
if we have a runtime assertions u0+s1 =10
then s1 got specialized to say 8, then when s1 will no longer be part of the graph we shall replace it with 8. 

now here is an interesting situation, if we have assert(s0==10) and prefer runtime assertions over guards ! 
what should happen in that case? we do want the assertion to be emitted, meaning that specializing s0 
should not lead to s0 being removed from the graph. 
Maybe the proper fix is that we should NEVER remove symbols used in runtime assertions from the graph ever?



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: specializing symbols after runtime assertions added cause codegen issue"


Fix https://github.com/pytorch/pytorch/issues/153650
if we have a runtime assertions u0+s1 =10
then s1 got specialized to say 8, then when s1 will no longer be part of the graph we shall replace it with 8. 

now here is an interesting situation, if we have assert(s0==10) and prefer runtime assertions over guards ! 
what should happen in that case? we do want the assertion to be emitted, meaning that specializing s0 
should not lead to s0 being removed from the graph. 
Maybe the proper fix is that we should NEVER remove symbols used in runtime assertions from the graph ever?



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: specializing symbols after runtime assertions added cause codegen issue"


Fix https://github.com/pytorch/pytorch/issues/153650
if we have a runtime assertions u0+s1 =10
then s1 got specialized to say 8, then when s1 will no longer be part of the graph we shall replace it with 8. 

now here is an interesting situation, if we have assert(s0==10) and prefer runtime assertions over guards ! 
what should happen in that case? we do want the assertion to be emitted, meaning that specializing s0 
should not lead to s0 being removed from the graph. 
Maybe the proper fix is that we should NEVER remove symbols used in runtime assertions from the graph ever?



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka commented:
change the fix strategy trying the new one.

laithsakka committed:
Update on "Fix: specializing symbols after runtime assertions added cause codegen issue"


Fix https://github.com/pytorch/pytorch/issues/153650
Two things:
1) input nodes that represents symbols that are used in runtime assertions should NOT be deleted. 
2) during codegen and runtime assertion insertion, do not apply the replacements to input nodes that represent symbols that are used in runtime assertions.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: specializing symbols after runtime assertions added cause codegen issue"


Fix https://github.com/pytorch/pytorch/issues/153650
Two things:
1) input nodes that represents symbols that are used in runtime assertions should NOT be deleted. 
2) during codegen and runtime assertion insertion, do not apply the replacements to input nodes that represent symbols that are used in runtime assertions.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: specializing symbols after runtime assertions added cause codegen issue"


Fix https://github.com/pytorch/pytorch/issues/153650
Two things:
1) input nodes that represents symbols that are used in runtime assertions should NOT be deleted. During remove_unused_inputs if the runtime assertions were inserted correctly and non missed this wont happpen. 

2) during codegen and runtime assertion insertion, do not apply the replacements to input nodes that represent symbols that are used in runtime assertions in order to avoid missing inserting runtime assertions depending on those symbols.
and to prevent (in the case of inductor) generating invalid code.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: Ban replacements for unbacked"


Fix https://github.com/pytorch/pytorch/issues/153650
Ok trying approach 3.
consider this example
```
torch.compile(dynamic=True)
def func(x, y, z):
        torch._check(x.size()[0]==y.size()[0])
        return x
x = torch.rand(4,4)
torch._dynamo.decorators.mark_unbacked(x, 0)

with fresh_inductor_cache():
        func(x, torch.rand(4,4), torch.tensor([1]))
```

we have a check u0==s0 and we add such replacement , this is problematic right now :
1) the assert is not inserted in insert_deferred_runtime_asserts
2) the assert is not codegen during inductor. 
3) 1, can also lead to some input of the graph being deleted while referenced in runtime assertion.

To fix this, we just ban replacements on unbacked. (except those that happen during propagation of unbacked symints, those are different). 

things I tried that wont work:
1) specializing during codegen, this wont fix 1,3. and also can lead to generating nop assertions in some scenarios (that should not be nops).
2) access original expr (instead of replaced expression).
 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: Ban replacements for unbacked"


Fix https://github.com/pytorch/pytorch/issues/153650
Ok trying approach 3.
consider this example
```
torch.compile(dynamic=True)
def func(x, y, z):
        torch._check(x.size()[0]==y.size()[0])
        return x
x = torch.rand(4,4)
torch._dynamo.decorators.mark_unbacked(x, 0)

with fresh_inductor_cache():
        func(x, torch.rand(4,4), torch.tensor([1]))
```

we have a check u0==s0 and we add such replacement , this is problematic right now :
1) the assert is not inserted in insert_deferred_runtime_asserts
2) the assert is not codegen during inductor. 
3) 1, can also lead to some input of the graph being deleted while referenced in runtime assertion.

To fix this, we just ban replacements on unbacked. (except those that happen during propagation of unbacked symints, those are different and ok).

other related issue: 
what if we add a check u0==s0 and later on we we specializes s0 to 1! in this case, right now we will not generate
the assert for that during insert_deferred_runtime_asserts because we never see s0 and we will generate a code
that reference an non-existent symbol.
one solution is to to apply the replacements on the runtime assert expressions. with out banning replacements on 
unbacked this is not Ok, with banning this become trivial, we apply replacements on the rsa expression. 
In fact this expression can simply be a symnode now. instead of sympy expression.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
HERE the problems with banning unbacked replacements:
1. what about when we defer guards to be runtime assertions for backed, 
 func(s1, s2) 
      torch.check(s1==s2) --->
with replacements args become 
func(ss, s2) 
      torch.check(s1==s2) 
      
2. Problem two, it effects quality of reasoning ! 
(shall we add axioms instead?)

laithsakka committed:
Update on "Fix: Replacements can cause runtime assertions to disappear or can cause invalid inductor code.  "


Lets explore firs a couple of problem related to replacements and runtime assertions.

## example problem 1
if we have a runtime assertions u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added,  the function f(u0, s0) will become f(s0, s0), this leads to the assert to not be inserted in during insert_deferred_runtime_asserts. since insert_deferred_runtime_asserts logic insert each assertions once all its input are seen but u0 will never bee seen.
same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

## example problem 2
consider u0==s0, where u0 is coming from a call to .item(), in that case insert_deferred_runtime_asserts will see it and it to the graph with no issues. Imagine later on that a specialization happen to s0 to become 2. 

... to be continued.


Other approaches that did not work :
### ban replacements on unbacked. 
1) does not work when we defer runtime assertions on backed ! s0==s1. we could also ban such replacements
but problem 2 become more problematic.  
2) Problem two, it effects quality of reasoning ! in a bad way. 

### Apply specialization on runtime assertions before codegen .
1) can fix some inductor issues, but may lead also to runtime assertions becoming NOPs also sometimes.
2) does not fix the issue of not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected. 
 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: Replacements can cause runtime assertions to disappear and can cause invalid inductor code.  "


Lets explore firs a couple of problem related to replacements and runtime assertions.

## example problem 1
if we have a runtime assertions u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added,  the function f(u0, s0) will become f(s0, s0), this leads to the assert to not be inserted in during insert_deferred_runtime_asserts. since insert_deferred_runtime_asserts logic insert each assertions once all its input are seen but u0 will never bee seen.
same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

## example problem 2
consider u0==s0, where u0 is coming from a call to .item() Imagine later on that a specialization happen to s0 to become 2.  in that case s0 as input wont be seen during insert_deferred_runtime_asserts and the assertion wont be inserted in the graph. Worse, inductor will generate some code that refer to s0 in the cpp wrapper while it does not exisit, causing a failure. 
see xref: 


Other approaches that did not work :
### ban replacements on unbacked. 
1) does not work when we defer runtime assertions on backed ! s0==s1. we could also ban such replacements
but problem 2 become more problematic.  
2) Problem two, it effects quality of reasoning ! in a bad way. 

### Apply specialization on runtime assertions before codegen .
1) can fix some issues, but may lead also to runtime assertions becoming NOPs. 
2) does not fix the issue if not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected. 
 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: Replacements can cause runtime assertions to disappear and can cause invalid inductor code.  "


Lets explore firs a couple of problem related to replacements and runtime assertions.

## example problem 1
if we have a runtime assertions u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added,  the function f(u0, s0) will become f(s0, s0), this leads to the assert to not be inserted in during insert_deferred_runtime_asserts. since insert_deferred_runtime_asserts logic insert each assertions once all its input are seen but u0 will never bee seen.
same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

## example problem 2
consider u0==s0, where u0 is coming from a call to .item() Imagine later on that a specialization happen to s0 to become 2.  in that case s0 as input wont be seen during insert_deferred_runtime_asserts and the assertion wont be inserted in the graph. Worse, inductor will generate some code that refer to s0 in the cpp wrapper while it does not exisit, causing a failure. 
see xref: 


Other approaches that did not work :
### ban replacements on unbacked. 
1) does not work when we defer runtime assertions on backed ! s0==s1. we could also ban such replacements
but problem 2 become more problematic.  
2) Problem two, it effects quality of reasoning ! in a bad way. 

### Apply specialization on runtime assertions before codegen .
1) can fix some issues, but may lead also to runtime assertions becoming NOPs. 
2) does not fix the issue if not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected. 
 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bobrenjc93 commented on test/test_dynamic_shapes.py:
```diff
@@ -3017,9 +3027,44 @@ def func(a):
         with self.assertRaises(RuntimeError):
             func(torch.tensor([5]))
 
+    # Test a situation where we generate a runtime assert i.e: u1==s1, then we spcialize s1
```
s/spcialize/specialize

jansel reviewed (changes requested):
test failures?

laithsakka commented:
> test failures?

those are just for a new API added i will make it private.

laithsakka committed:
Update on "Fix: Replacements can cause runtime assertions to disappear and can cause invalid inductor code.  "


Lets explore firs a couple of problem related to replacements and runtime assertions.

#### example problem 1
if we have a runtime assertions that u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added, the function f(u0, s0) will become f(s0, s0), this leads to the assert  not being inserted during insert_deferred_runtime_asserts. 
The reason is that insert_deferred_runtime_asserts logic insert each assertion once all its inputs are seen,  but u0 will never be seen. Same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

#### example problem 2
Consider u0==s0, where u0 is coming from a call to .item() Imagine later on that a specialization happens to s0 to become 2. In that case s0 as input wont be seen during insert_deferred_runtime_asserts and the assertion won't be inserted in the graph. Worse, Inductor will generate some code that refers to s0 in the cpp wrapper while it does not exist, causing a failure.
internal xref: https://fb.workplace.com/groups/1075192433118967/permalink/1669766396994898/

## The solution :
Runtime assertions insertion loops depend on detecting that the symbols that are used in the runtime assertions are seen, note that those symbols are either graph inputs or generated in the graph from data dependent ops like .item().


The issues above happen when symbols are graph inputs, in order to force the symbols to exist in the graph and to be seen by the runtime assertions we do not do replacements on placeholders expressions during codegen and during runtime assertions insertion.

This should not have performance overhead, since we already optimized the graph with replacements, the only effect is not mistakenly dropping graph inputs that are used in runtime assertions.
I added extended testing. A solo unrelated follow up that I noticed, is that we might want to rename unbacked symbols in runtime assertions when we do unbacked renaming, but that's a different issue.

Other approaches that did not work :
#### ban replacements on unbacked.
1. does not work when we defer runtime assertions on backed ex: s0==s1. we could also ban such replacements
but problem 2 becomes more problematic.
2. Problem two, it affects the quality of reasoning ! in a bad way.

#### Apply specialization on runtime assertions before codegen .
1. Can fix some issues, but may lead also to runtime assertions becoming NOPs.
2. Does not fix the issue if not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected.



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
address benign failures.

laithsakka committed:
Update on "Fix: Replacements can cause runtime assertions to disappear and can cause invalid inductor code.  "


Lets explore firs a couple of problem related to replacements and runtime assertions.

#### example problem 1
if we have a runtime assertions that u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added, the function f(u0, s0) will become f(s0, s0), this leads to the assert  not being inserted during insert_deferred_runtime_asserts. 
The reason is that insert_deferred_runtime_asserts logic insert each assertion once all its inputs are seen,  but u0 will never be seen. Same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

#### example problem 2
Consider u0==s0, where u0 is coming from a call to .item() Imagine later on that a specialization happens to s0 to become 2. In that case s0 as input wont be seen during insert_deferred_runtime_asserts and the assertion won't be inserted in the graph. Worse, Inductor will generate some code that refers to s0 in the cpp wrapper while it does not exist, causing a failure.
internal xref: https://fb.workplace.com/groups/1075192433118967/permalink/1669766396994898/

## The solution :
Runtime assertions insertion loops depend on detecting that the symbols that are used in the runtime assertions are seen, note that those symbols are either graph inputs or generated in the graph from data dependent ops like .item().


The issues above happen when symbols are graph inputs, in order to force the symbols to exist in the graph and to be seen by the runtime assertions we do not do replacements on placeholders expressions during codegen and during runtime assertions insertion.

This should not have performance overhead, since we already optimized the graph with replacements, the only effect is not mistakenly dropping graph inputs that are used in runtime assertions.
I added extended testing. A solo unrelated follow up that I noticed, is that we might want to rename unbacked symbols in runtime assertions when we do unbacked renaming, but that's a different issue.

Other approaches that did not work :
#### ban replacements on unbacked.
1. does not work when we defer runtime assertions on backed ex: s0==s1. we could also ban such replacements
but problem 2 becomes more problematic.
2. Problem two, it affects the quality of reasoning ! in a bad way.

#### Apply specialization on runtime assertions before codegen .
1. Can fix some issues, but may lead also to runtime assertions becoming NOPs.
2. Does not fix the issue if not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected.



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: Replacements can cause runtime assertions to disappear and can cause invalid inductor code.  "


Lets explore firs a couple of problem related to replacements and runtime assertions.

#### example problem 1
if we have a runtime assertions that u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added, the function f(u0, s0) will become f(s0, s0), this leads to the assert  not being inserted during insert_deferred_runtime_asserts. 
The reason is that insert_deferred_runtime_asserts logic insert each assertion once all its inputs are seen,  but u0 will never be seen. Same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

#### example problem 2
Consider u0==s0, where u0 is coming from a call to .item() Imagine later on that a specialization happens to s0 to become 2. In that case s0 as input wont be seen during insert_deferred_runtime_asserts and the assertion won't be inserted in the graph. Worse, Inductor will generate some code that refers to s0 in the cpp wrapper while it does not exist, causing a failure.
internal xref: https://fb.workplace.com/groups/1075192433118967/permalink/1669766396994898/

## The solution :
Runtime assertions insertion loops depend on detecting that the symbols that are used in the runtime assertions are seen, note that those symbols are either graph inputs or generated in the graph from data dependent ops like .item().


The issues above happen when symbols are graph inputs, in order to force the symbols to exist in the graph and to be seen by the runtime assertions we do not do replacements on placeholders expressions during codegen and during runtime assertions insertion.

This should not have performance overhead, since we already optimized the graph with replacements, the only effect is not mistakenly dropping graph inputs that are used in runtime assertions.
I added extended testing. A solo unrelated follow up that I noticed, is that we might want to rename unbacked symbols in runtime assertions when we do unbacked renaming, but that's a different issue.

Other approaches that did not work :
#### ban replacements on unbacked.
1. does not work when we defer runtime assertions on backed ex: s0==s1. we could also ban such replacements
but problem 2 becomes more problematic.
2. Problem two, it affects the quality of reasoning ! in a bad way.

#### Apply specialization on runtime assertions before codegen .
1. Can fix some issues, but may lead also to runtime assertions becoming NOPs.
2. Does not fix the issue if not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected.



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchbot merge -f

laithsakka committed:
Update on "Fix: Replacements can cause runtime assertions to disappear and can cause invalid inductor code.  "


Lets explore firs a couple of problem related to replacements and runtime assertions.

#### example problem 1
if we have a runtime assertions that u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added, the function f(u0, s0) will become f(s0, s0), this leads to the assert  not being inserted during insert_deferred_runtime_asserts. 
The reason is that insert_deferred_runtime_asserts logic insert each assertion once all its inputs are seen,  but u0 will never be seen. Same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

#### example problem 2
Consider u0==s0, where u0 is coming from a call to .item() Imagine later on that a specialization happens to s0 to become 2. In that case s0 as input wont be seen during insert_deferred_runtime_asserts and the assertion won't be inserted in the graph. Worse, Inductor will generate some code that refers to s0 in the cpp wrapper while it does not exist, causing a failure.
internal xref: https://fb.workplace.com/groups/1075192433118967/permalink/1669766396994898/

## The solution :
Runtime assertions insertion loops depend on detecting that the symbols that are used in the runtime assertions are seen, note that those symbols are either graph inputs or generated in the graph from data dependent ops like .item().


The issues above happen when symbols are graph inputs, in order to force the symbols to exist in the graph and to be seen by the runtime assertions we do not do replacements on placeholders expressions during codegen and during runtime assertions insertion.

This should not have performance overhead, since we already optimized the graph with replacements, the only effect is not mistakenly dropping graph inputs that are used in runtime assertions.
I added extended testing. A solo unrelated follow up that I noticed, is that we might want to rename unbacked symbols in runtime assertions when we do unbacked renaming, but that's a different issue.

Other approaches that did not work :
#### ban replacements on unbacked.
1. does not work when we defer runtime assertions on backed ex: s0==s1. we could also ban such replacements
but problem 2 becomes more problematic.
2. Problem two, it affects the quality of reasoning ! in a bad way.

#### Apply specialization on runtime assertions before codegen .
1. Can fix some issues, but may lead also to runtime assertions becoming NOPs.
2. Does not fix the issue if not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected.



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
address faluire

laithsakka committed:
Update on "Fix: Replacements can cause runtime assertions to disappear and can cause invalid inductor code.  "


Lets explore firs a couple of problem related to replacements and runtime assertions.

#### example problem 1
if we have a runtime assertions that u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added, the function f(u0, s0) will become f(s0, s0), this leads to the assert  not being inserted during insert_deferred_runtime_asserts. 
The reason is that insert_deferred_runtime_asserts logic insert each assertion once all its inputs are seen,  but u0 will never be seen. Same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

#### example problem 2
Consider u0==s0, where u0 is coming from a call to .item() Imagine later on that a specialization happens to s0 to become 2. In that case s0 as input wont be seen during insert_deferred_runtime_asserts and the assertion won't be inserted in the graph. Worse, Inductor will generate some code that refers to s0 in the cpp wrapper while it does not exist, causing a failure.
internal xref: https://fb.workplace.com/groups/1075192433118967/permalink/1669766396994898/

## The solution :
Runtime assertions insertion loops depend on detecting that the symbols that are used in the runtime assertions are seen, note that those symbols are either graph inputs or generated in the graph from data dependent ops like .item().


The issues above happen when symbols are graph inputs, in order to force the symbols to exist in the graph and to be seen by the runtime assertions we do not do replacements on placeholders expressions during codegen and during runtime assertions insertion.

This should not have performance overhead, since we already optimized the graph with replacements, the only effect is not mistakenly dropping graph inputs that are used in runtime assertions.
I added extended testing. A solo unrelated follow up that I noticed, is that we might want to rename unbacked symbols in runtime assertions when we do unbacked renaming, but that's a different issue.

Other approaches that did not work :
#### ban replacements on unbacked.
1. does not work when we defer runtime assertions on backed ex: s0==s1. we could also ban such replacements
but problem 2 becomes more problematic.
2. Problem two, it affects the quality of reasoning ! in a bad way.

#### Apply specialization on runtime assertions before codegen .
1. Can fix some issues, but may lead also to runtime assertions becoming NOPs.
2. Does not fix the issue if not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected.



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Fix: Replacements can cause runtime assertions to disappear and can cause invalid inductor code.  "


Lets explore firs a couple of problem related to replacements and runtime assertions.

#### example problem 1
if we have a runtime assertions that u0==s0, u0 is an input coming from mark_unbacked. A replacement u0=s0 will be added, the function f(u0, s0) will become f(s0, s0), this leads to the assert  not being inserted during insert_deferred_runtime_asserts. 
The reason is that insert_deferred_runtime_asserts logic insert each assertion once all its inputs are seen,  but u0 will never be seen. Same thing can happen when we defer assertion on backed i.e: s0==s2 ..etc.

#### example problem 2
Consider u0==s0, where u0 is coming from a call to .item() Imagine later on that a specialization happens to s0 to become 2. In that case s0 as input wont be seen during insert_deferred_runtime_asserts and the assertion won't be inserted in the graph. Worse, Inductor will generate some code that refers to s0 in the cpp wrapper while it does not exist, causing a failure.
internal xref: https://fb.workplace.com/groups/1075192433118967/permalink/1669766396994898/

## The solution :
Runtime assertions insertion loops depend on detecting that the symbols that are used in the runtime assertions are seen, note that those symbols are either graph inputs or generated in the graph from data dependent ops like .item().


The issues above happen when symbols are graph inputs, in order to force the symbols to exist in the graph and to be seen by the runtime assertions we do not do replacements on placeholders expressions during codegen and during runtime assertions insertion.

This should not have performance overhead, since we already optimized the graph with replacements, the only effect is not mistakenly dropping graph inputs that are used in runtime assertions.
I added extended testing. A solo unrelated follow up that I noticed, is that we might want to rename unbacked symbols in runtime assertions when we do unbacked renaming, but that's a different issue.

Other approaches that did not work :
#### ban replacements on unbacked.
1. does not work when we defer runtime assertions on backed ex: s0==s1. we could also ban such replacements
but problem 2 becomes more problematic.
2. Problem two, it affects the quality of reasoning ! in a bad way.

#### Apply specialization on runtime assertions before codegen .
1. Can fix some issues, but may lead also to runtime assertions becoming NOPs.
2. Does not fix the issue if not inserting runtime assertions during insert_deferred_runtime_asserts due to input not being detected.



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchbot revert
